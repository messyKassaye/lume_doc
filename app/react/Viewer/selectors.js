import { createSelector } from 'reselect';
import { fromJS as Immutable } from 'immutable';

const documentViewer = s => s.documentViewer;

function isRelationshipAReference(doc, reference) {
  return (
    reference.get('entity') === doc.get('sharedId') &&
    typeof reference.getIn(['reference', 'text']) !== 'undefined'
  );
}

const parseReferences = (doc, refs) =>
  refs
    .filter(r => isRelationshipAReference(doc, r))
    .reduce((hubs, r) => {
      if (hubs.indexOf(r.get('hub')) === -1) {
        hubs.push(r.get('hub'));
      }
      return hubs;
    }, [])
    .reduce((memo, hubId) => {
      let references = memo;
      const baseReference = refs.find(
        r => isRelationshipAReference(doc, r) && r.get('hub') === hubId
      );
      const otherRelationshipsOfHub = refs.filter(
        r => r.get('_id') !== baseReference.get('_id') && r.get('hub') === hubId
      );
      otherRelationshipsOfHub.forEach(r => {
        references = references.push(baseReference.set('associatedRelationship', r));
      });
      return references;
    }, Immutable([]));

const selectDoc = createSelector(
  s => documentViewer(s).doc,
  doc => doc
);
const selectRefs = createSelector(
  s => documentViewer(s).references,
  references => references
);

const selectReferences = createSelector(selectDoc, selectRefs, parseReferences);

const selectTargetDoc = createSelector(
  s => documentViewer(s).targetDoc,
  doc => doc
);

const selectTargetRefs = createSelector(
  s => documentViewer(s).targetDocReferences,
  references => references
);

const selectTargetReferences = createSelector(selectTargetDoc, selectTargetRefs, parseReferences);

export { selectDoc, selectReferences, selectTargetDoc, selectTargetReferences, parseReferences };
