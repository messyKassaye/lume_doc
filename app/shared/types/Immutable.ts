import Immutable from 'immutable';

export type IImmutable<T> = T extends string
  ? string
  : T extends number
  ? number
  : T extends boolean
  ? boolean
  : T extends null
  ? null
  : T extends undefined
  ? undefined
  : T extends Array<infer Elem>
  ? Immutable.List<IImmutable<Elem>>
  : {
      toJS(): T;
      get<Field extends keyof T>(_field: Field): IImmutable<T[Field]>;
    };
