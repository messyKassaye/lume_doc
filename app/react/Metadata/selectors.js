import { createSelector } from 'reselect';
import { entityDefaultDocument } from 'shared/entityDefaultDocument';

import Immutable from 'immutable';
import formater from './helpers/formater';

const indexValues = t =>
  t.set(
    'values',
    t.get('values').reduce((indexed, value) => {
      if (value.get('values')) {
        return indexed.merge(indexValues(value).get('values'));
      }
      return indexed.set(value.get('id'), value);
    }, new Immutable.Map({}))
  );

const indexedThesaurus = createSelector(
  s => s.thesauris,
  thesaurus => thesaurus.map(t => indexValues(t))
);

const formatMetadata = createSelector(
  s => s.templates,
  indexedThesaurus,
  s => s.settings,
  (_s, doc, sortProperty, references, options) => ({ doc, sortProperty, references, options }),
  (templates, thesauris, settings, { doc, sortProperty, references, options }) => {
    const defaultDoc = entityDefaultDocument(
      doc.documents,
      doc.language,
      settings
        .get('languages')
        .find(l => l.get('default'))
        .get('key')
    );

    if (sortProperty) {
      return formater.prepareMetadataForCard(
        Object.assign(doc, { defaultDoc }),
        templates,
        thesauris,
        sortProperty
      ).metadata;
    }

    return formater.prepareMetadata(
      Object.assign(doc, { defaultDoc }),
      templates,
      thesauris,
      references,
      options
    ).metadata;
  }
);

const metadataSelectors = {
  formatMetadata,
  indexedThesaurus,
};

export { metadataSelectors };
