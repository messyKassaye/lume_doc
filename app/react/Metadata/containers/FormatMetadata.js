import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import React from 'react';
import Immutable from 'immutable';
import { metadataSelectors } from '../selectors';
import Metadata from '../components/Metadata';

const removeUneededProps = ({ templates, thesauris, settings, excludePreview, ...rest }) => rest;

const BaseFormatMetadata = ({
  additionalMetadata,
  sortedProperty,
  entity,
  relationships,
  ...props
}) => (
  <Metadata
    metadata={additionalMetadata.concat(
      metadataSelectors.formatMetadata(props, entity, sortedProperty, relationships, {
        excludePreview: props.excludePreview,
      })
    )}
    compact={!!sortedProperty}
    {...removeUneededProps(props)}
  />
);

BaseFormatMetadata.defaultProps = {
  sortedProperty: '',
  additionalMetadata: [],
  relationships: Immutable.fromJS([]),
  excludePreview: false,
};

BaseFormatMetadata.propTypes = {
  entity: PropTypes.shape({
    metadata: PropTypes.object,
  }).isRequired,
  relationships: PropTypes.object,
  additionalMetadata: PropTypes.arrayOf(
    PropTypes.shape({
      label: PropTypes.string,
      value: PropTypes.oneOfType([
        PropTypes.string,
        PropTypes.number,
        PropTypes.arrayOf(PropTypes.string),
        PropTypes.arrayOf(
          PropTypes.shape({
            value: PropTypes.string,
          })
        ),
      ]),
    })
  ),
  sortedProperty: PropTypes.string,
  excludePreview: PropTypes.bool,
};

export function mapStateToProps(state, { entity, sortedProperty = '' }) {
  return {
    templates: state.templates,
    thesauris: state.thesauris,
    settings: state.settings.collection,
    entity,
    sortedProperty,
  };
}

export const FormatMetadata = connect(mapStateToProps)(BaseFormatMetadata);
