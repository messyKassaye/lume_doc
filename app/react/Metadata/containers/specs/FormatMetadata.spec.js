import React from 'react';
import { shallow } from 'enzyme';
import Immutable from 'immutable';

import { FormatMetadata } from '../FormatMetadata';
import { metadataSelectors } from '../../selectors';

describe('FormatMetadata', () => {
  it('should render Metadata component passing the formatted metadata', () => {
    spyOn(metadataSelectors, 'formatMetadata').and.returnValue([{ formated: 'metadata' }]);
    const props = {
      templates: [],
      thesauris: [],
      entity: {},
      sortedProperty: 'sortedProperty',
    };
    const component = shallow(<FormatMetadata.WrappedComponent {...props} />);
    expect(component).toMatchSnapshot();
  });

  it('should unshift additional metadata if passed', () => {
    const props = {
      templates: Immutable.fromJS([
        {
          _id: 'template',
          properties: [{ name: 'preview', type: 'preview' }],
        },
      ]),
      thesauris: Immutable.fromJS([{ _id: 'thesauris', values: [] }]),
      settings: Immutable.fromJS({ languages: [{ key: 'es', default: true }] }),
      entity: {
        template: 'template',
        documents: [{ _id: 'docId' }],
      },
      additionalMetadata: [{ more: 'data' }, { and: 'more' }],
    };

    const component = shallow(<FormatMetadata.WrappedComponent {...props} />);
    expect(component).toMatchSnapshot();
  });
});
