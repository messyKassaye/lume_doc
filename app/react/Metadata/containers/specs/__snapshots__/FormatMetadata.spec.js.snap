// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormatMetadata should render Metadata component passing the formatted metadata 1`] = `
<Metadata
  compact={true}
  groupGeolocations={false}
  highlight={Array []}
  metadata={
    Array [
      Object {
        "formated": "metadata",
      },
    ]
  }
  renderLabel={[Function]}
/>
`;

exports[`FormatMetadata should unshift additional metadata if passed 1`] = `
<Metadata
  compact={false}
  groupGeolocations={false}
  highlight={Array []}
  metadata={
    Array [
      Object {
        "more": "data",
      },
      Object {
        "and": "more",
      },
      Object {
        "indexInTemplate": 0,
        "label": undefined,
        "name": "preview",
        "noLabel": false,
        "style": "contain",
        "translateContext": "template",
        "type": "image",
        "value": "/api/files/docId.jpg",
      },
    ]
  }
  renderLabel={[Function]}
/>
`;
